Changes to solution description (#2)
duftler/click-to-deploy: k8s/spinnaker/README.md

> 

Reviewer: Where does this readme appear? Does this show up in the Marketplace with the solution, or just in the click-to-deploy repo? Reason I ask is I'm wondering why we have a sentence, and a level-2 heading, describing Google Click to Deploy if this readme is for Spinnaker specifically.

Reviewer: should match other click-to-deploy k8s apps

Author: Hm, it looks like what's in the text files may actually not be what's in the Marketplace UI, see e.g. https://github.com/duftler/click-to-deploy/blob/master/k8s/nginx/README.md#about-google-click-to-deploy

Will leave as-is, in that case.

Popular open stacks on Kubernetes packaged by Google.

## About Spinnaker for Google Cloud Platform

> This solution installs a single instance of Spinnaker into a GKE cluster in a

Reviewer: s/into/onto/ 
I think.

Author: @duftler Thoughts? I'm happy either way

Reviewer 2: I _think_ 'into' reads better, but now I've said both words so many times I can't tell anymore.

production-ready configuration. The installation follows recommended practices
for running Spinnaker on Google Cloud Platform, and is integrated with related

> Google Cloud services, such as [Cloud Build](https://cloud.google.com/cloud-build/).

Reviewer 2: At the moment, we don't have a GCB integration. Once 1.14 is out, and we bump the referenced version to 1.14, we will.

Reviewer 2: Actually, not sure my comment is accurate.

Reviewer 2: I was originally referring to the GCB stage, but I'm sure we can pre-configure the GCB pub/sub account (we've done this for GCR).

Author: That's what I had in mind here too. There's non-zero things we can already do to to integrate Spinnaker with Cloud Build, such as enabling triggering from a Cloud Build job. If we're not currently doing *any* such setup, I can replace this with GCR.

This solution also provides a simplified configuration experience, as well as a
management environment and workflow for ongoing administration of Spinnaker.

## Architecture

### TODO: Create architecture diagram

![Architecture diagram](resources/spinnaker-k8s-app-architecture.png)

Spinnaker is composed of a number of individual
[microservices](https://www.spinnaker.io/reference/architecture/). These
will be deployed in their own Kuberetes Pods, managed by Deployment objects,
behind Service objects.

[Halyard](https://www.spinnaker.io/reference/halyard/) is Spinnaker's
configuration service and consists of a CLI and a daemon. The CLI will be
installed in the management environment (based on Cloud Shell) included in
this solution. The daemon will be installed in a Pod managed by a StatefulSet
object.

> [Spin](https://www.spinnaker.io/guides/spin/app/) is Spinnaker's CLI. It is also

Reviewer: We link above to the page on spinnaker.io that describes all the microservices, and here we call out two of them specifically. I suggest adding a sentence above this line that says something like "Two of these microservices are important for our puposes here." And then set off Halyard and Spin with bullets.

Reviewer: Instead, add, below that first para, "We provide the following two components to help you manage your Spinnaker instance.

available in the management environment.

> 

Reviewer: ...then add a heading here. 

## Accessing Spinnaker securely

Author: Modified slightly since we're talking about the Architecture of the setup we've created here. Please take a look!

> As a safe default, the Spinnaker instance is not exposed to external traffic, and

Reviewer: "...is not exposed to external traffic. It's accessed via port fowarding, which can be set up with..."

is accessed via port forwarding. Port forwarding can be set up with a single command

> from the management environment. Alternatively, the management environment allos you

Reviewer 2: s/allos/allows

to expose Spinnaker via the [Identity-Aware Proxy](https://cloud.google.com/iap/), using
a secure domain.

# Installation

## Quick install via the management console

You must use the management environment, based on [Cloud Shell](https://cloud.google.com/shell/),
to perform the installation.

[![Provision Spinnaker](https://gstatic.com/cloudssh/images/open-btn.png)](https://console.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https://github.com/GoogleCloudPlatform/click-to-deploy.git&cloudshell_working_dir=k8s/spinnaker/scripts/install&cloudshell_tutorial=provision-spinnaker.md)

### Prerequisites

#### Set up command line tools

The management environment has all the required tools pre-installed.

#### Create a Google Kubernetes Engine cluster

> By default, the installation process will create a new GKE cluster for your Spinnaker

Reviewer: s/will create/creates/

instance. To use an existing cluster, you must ensure that it has:

* IP aliases enabled (since we are using a
[Cloud Memorystore for Redis](https://cloud.google.com/memorystore/) instance)

> * `Full Cloud Platform` scope for its nodes (if using the default service account)

Reviewer: Do I need to perform any other tricks to point the script(s) at this pre-existing cluster?

Reviewer: @duftler 
Is there  a link to how to set the full cloud platform scope?

Reviewer 2: Just need to edit the cluster name in the generated properties file to use an existing cluster.

Users can set the scope for a GKE cluster's nodes either with an argument to gcloud (visible via the `--help` flag to gcloud) or via the pantheon ui (which has `Full Cloud Platform` scope as one of the choices in the pulldown).

#### Install Spinnaker for Google Cloud Platform

When you click on the "Open in Google Cloud Shell" button above, the management

> environment will open and ask you to copy/paste and execute a number of commands.

Reviewer: s/will open and ask/asks/.

Also, add another sentence: "Click to copy/paste, then press Enter to execute."

> As part of this process, a `properties` file will be generated. You can modify the

Reviewer: s/will be/is/

> values in the generated `properties` file to fine-tune the desired configuration

Reviewer: s/the desired/your/

> _prior_ to launching the actual `setup` script.

Reviewer: s/_prior_ to/_before_/

When the initial installation process completes, additional commands will be shown
to allow you to:

> * locate your Spinnaker deployment in the GKE Applications view

Reviewer: s/locate/find/

* access Spinnaker via port-forwarding
* expose Spinnaker securely using IAP

> * allow others to administer and/or access Spnnakeir

Reviewer 2: s/Spnnakeir/Spinnaker

and more.

#### Ongoing Management of Spinnaker

> To manage your Spinnaker instance once installed:

Reviewer: ```suggestion
To manage your Spinnaker instance once installed:
```
```suggestion
To manage your installed Spinnaker instance:
```

* Locate your Spinnaker installation by navigating to the newly-registered
Application in the
[GKE Applications](https://console.developers.google.com/kubernetes/application?project=_)
view.
* Click the link from the Spinnaker Application to open the management
environment in Cloud Shell, and update it by cloning this repo

> * Verify that you are managing the correct Spinnaker instance by ensuring that you

Reviewer: ```suggestion
* make sure you're
```

> are connected to the GKE cluster containing the Spinnaker instance you intend to manage

Reviewer: ```suggestion
connected to the GKE cluster containing the Spinnaker instance you want to manage.
```

Reviewer: Should we say how to to do this?

Reviewer: @duftler 
What's the command to do this?

> * Update your configuration but pulling the config from the Spinnaker instance into your

Reviewer 2: s/but/by

> management environment

Reviewer: ```suggestion
management environment.
```

> * Refresh the management environment interface to reflect your updated configuration

Reviewer: ```suggestion
* Refresh the management environment interface to reflect your updated configuration.
```

Reviewer: By refreshing the whole page/tab?

Reviewer: @demobox 
Add `update_console.sh` to this?

* Make all of your changes using `hal` in Cloud Shell, using the commands provided by the
management environment, or by entering commands directly. `hal` will validate and apply your

> changes

Reviewer: ```suggestion
changes.
```

Reviewer: Do we want another admonishment not to use `hal deploy connect`?

Reviewer: @duftler do we still want to tell folks not to use `hal deploy connect`?

* Push the updated configuration to your Spinnaker instance. The configuration is automatically

> backed up to GCS as part of this process

Reviewer: ```suggestion
backed up to GCS as part of this process.
```

* Apply your changes to the Spinnaker instance

Scripts are provided for each of these steps, and copy/paste commands and instructions
are rendered in the management environment in the Cloud Shell tutorial pane.

This approach allows an administrator to manage multiple Spinnaker instances from the
same management environment. The source of truth for each instance is always the instance
itself, with a secure backup in [Google Cloud Storage](https://cloud.google.com/storage).

> This makes it easy for multiple admins - each with their own management environment, to

Reviewer: ```suggestion
This makes it easy for multiple admins, each with their own management environment, to
```

Popular open stacks on Kubernetes packaged by Google.

## About Spinnaker for Google Cloud Platform

This solution installs a single instance of Spinnaker onto a GKE cluster in a
production-ready configuration. The installation follows recommended practices
for running Spinnaker on Google Cloud Platform, and is integrated with related
Google Cloud services, such as [Cloud Build](https://cloud.google.com/cloud-build/).

This solution also provides a simplified configuration experience, as well as a
management environment and workflow for ongoing administration of Spinnaker.

## Architecture

### TODO: Create architecture diagram

![Architecture diagram](resources/spinnaker-k8s-app-architecture.png)

Spinnaker is composed of a number of individual
[microservices](https://www.spinnaker.io/reference/architecture/). These
will be deployed in their own Kuberetes Pods, managed by Deployment objects,
behind Service objects.

> ### Management components

Author: @dorbin FYI

Reviewer: :+1:

We provide the following two components to help you manage your Spinnaker instance:

[Halyard](https://www.spinnaker.io/reference/halyard/) is Spinnaker's
configuration service and consists of a CLI and a daemon. The CLI will be
installed in the management environment (based on Cloud Shell) included in
this solution. The daemon will be installed in a Pod managed by a StatefulSet
object.

[Spin](https://www.spinnaker.io/guides/spin/app/) is Spinnaker's CLI. It is also
available in the management environment.

> ### Networking

Reviewer: ```suggestion
### Network access to Spinnaker
```

### Prerequisites

#### Set up command line tools

The management environment has all the required tools pre-installed.

#### Create a Google Kubernetes Engine cluster

By default, the installation process creates a new GKE cluster for your Spinnaker
instance. To use an existing cluster, you must ensure that it has:

* IP aliases enabled (since we are using a
[Cloud Memorystore for Redis](https://cloud.google.com/memorystore/) instance)
* `Full Cloud Platform` scope for its nodes (if using the default service account)

#### Install Spinnaker for Google Cloud Platform

When you click on the "Open in Google Cloud Shell" button above, the management
environment asks you to copy/paste and execute a number of commands. Click to copy/paste,
then press Enter to execute.

As part of this process, a `properties` file is generated. You can modify the
values in the generated `properties` file to fine-tune your configuration
_before_ launching the actual `setup` script.

When the initial installation process completes, additional commands will be shown
to allow you to:

* find your Spinnaker deployment in the GKE Applications view

> * access Spinnaker via port-forwarding

Author: @dorbin Where you referring to this place? Add something like "(you should use the provided script, not `hal deploy connect`)?

Reviewer: stet

* expose Spinnaker securely using IAP
* allow others to administer and/or access Spinnaker

and more.

#### Ongoing Management of Spinnaker

To manage your installed Spinnaker instance:

* Locate your Spinnaker installation by navigating to the newly-registered
Application in the
[GKE Applications](https://console.developers.google.com/kubernetes/application?project=_)
view.
* Click the link from the Spinnaker Application to open the management
environment in Cloud Shell, and update it by cloning this repo
* make sure you're
connected to the GKE cluster containing the Spinnaker instance you want to manage.
* Update your configuration by pulling the config from the Spinnaker instance into your
management environment.

> * Refresh the management environment interface to reflect your updated configuration.

Author: @dorbin Perhaps we should say "Update the..." here to avoid potential confusion that this implies reloading/refreshing the browser?

Reviewer: WFM

* Make all of your changes using `hal` in Cloud Shell, using the commands provided by the
management environment, or by entering commands directly. `hal` will validate and apply your
changes.
* Push the updated configuration to your Spinnaker instance. The configuration is automatically
backed up to GCS as part of this process.

> * Apply your changes to the Spinnaker instance

Author: @dorbin This might be another place where we caution users to use the provided script vs. calling `hal deploy apply`?
@duftler Is that correct?

Reviewer 2: If you think it's necessary, I can live with adding it here. But this is really just a list of the workflow steps, not the actual commands.
